ouroboros-consensus/src/Ouroboros/Consensus/Node/DbLock.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Ouroboros.Consensus.Node.DbLock (
+    DbLocked (..)
+  , withLockDB
+  , dbLockFile
+  ) where
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadTimer
+import           Data.Text
+import qualified Data.Time.Clock as Time
+import           System.FileLock
+
+import           Ouroboros.Consensus.Storage.FS.API
+import           Ouroboros.Consensus.Storage.FS.API.Types
+import           Ouroboros.Consensus.Util.IOLike
+
+-- We use an empty file as a lock of the db. Some systems may delete the empty
+-- file when all its handles are closed. This is not an issue, since the file is
+-- created if it doesn't exist.
+withLockDB :: HasFS IO h ->  FilePath -> IO a -> IO a
+withLockDB hasFS dbPath action = do
+    createDirectoryIfMissing hasFS True root
+    -- We want to avoid blocking the main thread at an ffi, to avoid
+    -- unresponsiveness. We shouldn't be tempted to use `withAsync` or anything
+    -- else which links the blocked thread to our main thread, since this would
+    -- negate the whole point of using async.
```

Comment:
I don't understand this. Do you mean unresponsiveness w.r.t. GC by the capability being blocked in an unsafe FFI call? Is `lockFile` an unsafe FFI call?

I also don't understand what you're saying about `withAsync`.

Reply:
unresponsiveness to ^C and timeouts. lockFile wraps an ffi call which is not interruptible and we don't want to have the main thread trapped in such a call. That's why I used async and let another thread do the actual call.

`withAsync` is usually mentioned as a better alternative to simple async (and this is what I firstly used) but it sends an async exception to the async thread and wait for it to terminate. This effectively blocks the main thread again and is what we tried to avoid in the first place.

Reply:
Ah, I see what you mean :+1:. Incorporate your response into the comment.